### ISSUE TITLE
'Include content' function fails data validation

### ISSUE THREAD

**Issue by VermaBSI** [ben-garside/XCP#114]
Items listed under PreXCP tab in Varun's CLAIMED list (23 in total) were added to enter at stage 20:10 or 10:19 but have failed to enter XCP at these stages. It appears the data it is looking for in each case is incomplete (a problem further upstream with Data Enrichment it seems?). The 10:19 entries were supposed to action a collation of files too. All these are sitting at stage 00:01 in PreXCP with pseudo pipeline allocation 96 (COMPLINAV).

**Comment by VermaBSI** [comment]
Sorry just remembered your email about collating files on inclusion so have moved this issue to INVALID list

**Comment by VermaBSI** [comment]
And now moving this back as a valid issue, in case the items set for manual inclusion to enter at 20.10 (involving no collation of files) should still have worked,,,

**Comment by ben-garside** [comment]
Could you please give an example of an item that should have worked please, just makes it easier to make sure I am looking at the right issue. 

In general, any item that fails DWH validation will fail ingest, this is based on these rules:

| field | required | validation |
| --- | --- | --- |
| projectStatus | 1 | NULL |
| standardsBody | 1 | NULL |
| originatingOrg | 1 | NULL |
| projectType | 1 | NULL |
| projectNumber | 1 | ^[0-9]{4}/[0-9]{5}$ |
| materialDescription | 1 | NULL |

So you can see in the `Feed Data` table on the item pages that you get this message in the `Validation Error` field:

```
[projectNumber] is a required field
[projectNumber] is not vlaid (^[0-9]{4}/[0-9]{5}$)
[projectType] is a required field
[projectStatus] is a required field
[standardsBody] is a required field
```

**Comment by VermaBSI** [comment]
Hi Ben,

Thanks for providing this information - it's confirmed what I suspected. 

First sorry you couldn't see any examples - but are you not able to see the 23 in the Pre-XCP CLAIMED list  under filter 'All Users', these are in my CLAIMED list, as I mentioned above? These are all the examples.

The problem is that when Data Enrichment create a record for exclusively third-party originating material (which does not become a Project for BSI publication), then a lot of the data fields listed above are not completed in SAP because they are N/A. 

Therefore would it be possible to not have these DWH validation rules for a UPI being manually included - or as a maximum, standardsBody and originatingOrg validation only once the UPI is matched in DWH?

**Comment by ben-garside** [comment]
The validation rules are global, this is because the material metadata from SAP(/DWH)  is being potentially used for multiple feeds/streams. So you could just not validate the data at all, but it would not validate for all items regardless of how the item is created or what stream they are in.

I think this issue should be closed as it is not an issue with the 1.3 build in UAT, and that you create a new issue (feature request) explaining how you would want to handle the validation going forward and we can talk about it as a group in our next ops meeting?

**Comment by VermaBSI** [comment]
Okay, not entirely clear why UPIs entering XCP through manual inclusion can't just be manually assigned a stream by a user as well like the feed is by the user, then data validation could be skipped as stream allocation has already been done -  so certainly one for discussion then :).

I'll close this then, and raise a new issue so we don't forget to have the conversation.

**Comment by ben-garside** [comment]
Because this was not the behaviour that was requested, It was requested that a user be able to add content to an existing feed, that is what it does.

For an item to be in XCP it _must_ have a valid UPI to link it up with BSI data.

**Comment by VermaBSI** [comment]
A UPI can be valid (i.e. record exists in DWH) but the record itself may not pass all the validation criteria set out above, right?

Yes manually add content to an existing feed as requested is fine, but now that I'm aware of all this validation criteria I am now suggesting that manually adding to an existing stream as well may be a useful bypass of this (?). We'll discuss it more with the rest.

**Comment by ben-garside** [comment]
The DWH data validation was an original requirement from the beginning of the project.

You could possibly bypass it as long as that UPI is never going to be used by any other item, I am not saying what is right or wrong, all I was doing was answering your question as to why it currently does not do this.

**Comment by VermaBSI** [comment]
Thanks Ben - just to clarify my communication (apologies), I was never asking why it _currently_ does not do this in my comment above about manually assigning a stream - I completely understand and get that DWH data validation is built into XCP as part of the original requirement to allocate streams. I am just now _newly_ suggesting that maybe being able to skip this is a new feature we  want specifically for UPIs entering the system through manual input. 

Your comment about validation rules being global gave the impression to me that it just wasn't possible without an entire re-think of data validation generally in XCP, that's why I suggested just being able to skip it, not re-engineer it - and in doing so I don't mean to imply this feature should have been provided originally :)
